libbeat/common/mapval/is_defs.go

```diff
@@ -115,20 +115,43 @@ var IsDuration = Is("is a duration", func(path Path, v interface{}) *Results {
 	)
 })
 
-// IsEqual tests that the given object is equal to the actual object.
-func IsEqual(to interface{}) IsDef {
-	return Is("equals", func(path Path, v interface{}) *Results {
-		if reflect.DeepEqual(v, to) {
+// IsEqualToTime ensures that the actual value is the given time, regardless of zone.
+func IsEqualToTime(to time.Time) IsDef {
+	return Is("equal to time", func(path Path, v interface{}) *Results {
+		actualTime, ok := v.(time.Time)
+		if !ok {
+			return SimpleResult(path, false, "Value %t was not a time.Time", v)
+		}
+
+		if actualTime.Equal(to) {
 			return ValidResult(path)
 		}
-		return SimpleResult(
-			path,
-			false,
-			fmt.Sprintf("objects not equal: actual(%v) != expected(%v)", v, to),
-		)
+
+		return SimpleResult(path, false, "actual(%v) != expected(%v)", actualTime, to)
 	})
 }
 
+// IsEqual tests that the given object is equal to the actual object.
+func IsEqual(to interface{}) IsDef {
+	// We use reflect.DeepEqual as a generic equality check, other types may need something
+	// more specific
+	switch converted := to.(type) {
+	case time.Time:
+		return IsEqualToTime(converted)
+	default:
+		return Is("equals", func(path Path, v interface{}) *Results {
+			if reflect.DeepEqual(v, to) {
+				return ValidResult(path)
+			}
+			return SimpleResult(
+				path,
+				false,
+				fmt.Sprintf("objects not equal: actual(%v) != expected(%v)", v, to),
+			)
+		})
+	}
+}
```

Comment:
As you are already a heavy user of reflection, how about this:

```
var equalChecks = map[reflect.Type]reflect.Value

// accepts and register equal function for function signatures:
// func(a, b T) bool
func RegisterEqual(fn interface{}) {
  v := reflect.ValueOf(fn)
  t := v.Type()
  if t.Kind() != reflect.Function { ... }
  if t.NumIn() != 2 { ...}
  if t.NumOut() != 1 { ... }
  if t.In(0) != t.In(1) { ... }
  if t.Out(0) != <bool> { ... }

  equalChecks[t.In(0)] = v
}

func IsEqual(to interface{}) IsDef {
  vTo := reflect.ValueOf(to)
  return Is("equals", func(path Path, in interface{}) *Results {
    vIn := reflect.ValueOf(in)
    if vTo.Type() != vIn.Type() { ... }

    // we always query, in case someone did add a comparison function after
    checker := equalChecks(vTo.Type())
    if checker == nil {
      // use DeepEqual
    }

    equal := checker.Call([]reflect.Value{vTo, VIn}).Bool()
    if !equal {
      ...
    }
  }
}
```

Reply:
This is pretty nifty, out of curiosity I wonder if using reflect is faster than a type switch? I assume they must be pretty similar perf-wise under the covers?

I am however worried about YAGNI. How many special equals cases are there? `reflect.DeepEquals` works pretty well, this is the first case it hasn't.

If there are only a few, I personally prefer the switch because its very concise, however, if we can foresee adding 10 of these cases a map sounds better.

WDYT?

Reply:
reflection is somewhat smaller.

In go an `interface{}` value has the internal representation:

```
type ifcType struct {
  typ *typeDescr
  value *unsafe.Pointer
}
```

The reflection package defines Value to be:

```
type Value struct {
  typ *typeDescr
  value *unsafe.Pointer
  flags flags
}
```

when using a type-switch, the generated code checks if the 'typ' matches and casts the ifc pointer to your appropriate type. If your target type is also a pointer, its a comparison + copy pointer into new variable.

When using reflection, each call to any Value method checks flags, types and so on, checking if your call is actually valid. The `Value.Call` method is super involved. Reflection is definitely slower, but gives you some flexibility here.

reflect.DeepEquals has it's limits. It compares each field (including private fields) in a struct. Private fields might be different, while 2 objects might still compare equal. E.g. when comparing regular expressions. DeepEquals compares public and private fields. But the regex package introduced an object pool pointer in tip. Each regexp.Regex type holds a unique pointer => comparison fails, if one does not compare the same pointer, even if two regular expression are generated from the same string.

Reply:
Cool, thanks for the detailed walkthrough. I think you're right, we'll probably add more of these, I'll add the registry.